### PR TITLE
Improve fetching and prefetching logic

### PR DESCRIPTION
**Pull request by aleDsz**
- **Closes** #124

### CODE REVIEW

File: test/goth_test.exs [hunk 1]
```diff
@@ -26,6 +26,35 @@ defmodule GothTest do
...
+    config = [
+      name: test,
+      source: {:service_account, random_service_account_credentials(), url: "http://localhost:#{bypass.port}"},
+      async: true
+    ]
+
+    start_supervised!({Goth, config})
+
+    # higher timeouts since calculating JWT is expensive
+    Process.sleep(1000)
```
Comment (aleDsz) [inline test/goth_test.exs:47]: I still doesn't know how to improve this check, tried with `pong` and still fails with `Goth.fetch/1` below

Comment (wojtekmach) [reply test/goth_test.exs:47]: handle_continue is always executed after the server already started so yeah, start_supervised cannot wait.

File: test/goth_test.exs [hunk 2]
```diff
@@ -26,6 +26,35 @@ defmodule GothTest do
     {:ok, ^token} = Goth.fetch(test)
   end
 
+  test "async fetch", %{test: test} do
```
Comment (wojtekmach) [inline test/goth_test.exs:29]:
```suggestion
  test "async prefetch", %{test: test} do
```

File: lib/goth/server.ex [hunk 3]
```diff
@@ -49,8 +32,25 @@ defmodule Goth.Server do
         Goth.HTTPClient.init({module, opts})
       end)
 
+    {async, opts} = Keyword.pop(opts, :async)
```
Comment (wojtekmach) [inline lib/goth/server.ex:35]: the option should be called `:prefetch` and the possible values are `:sync` and `:async`. Please also update Goth.start_link/1 docs accordingly.

File: lib/goth/server.ex [hunk 4]
```diff
@@ -49,8 +32,25 @@ defmodule Goth.Server do
...
+
     state = struct!(__MODULE__, opts)
 
+    if async do
+      {:ok, state, {:continue, :async_fetch}}
```
Comment (wojtekmach) [inline lib/goth/server.ex:40]: let's use the name _prefetch_

```suggestion
      {:ok, state, {:continue, :async_prefetch}}
```

File: lib/goth/server.ex [hunk 5]
```diff
@@ -57,18 +50,49 @@ defmodule Goth.Server do
...
+    end
+  end
+
+  @impl true
+  def handle_continue(:async_prefetch, state) do
+    do_prefetch(state)
+    {:noreply, state}
+  end
+
+  defp do_prefetch(state) do
```
Comment (wojtekmach) [inline lib/goth/server.ex:69]: let's skip the `do_`:

```suggestion
  defp prefetch(state) do
```

File: lib/goth/server.ex [hunk 6]
```diff
@@ -57,18 +50,49 @@ defmodule Goth.Server do
...
-    {:ok, state}
+  defp fetch_and_schedule_refresh(state) do
+    case Token.fetch(state) do
+      {:ok, token} ->
+        store_and_schedule_refresh(state, token)
+        {:ok, token}
+
+      {:error, _} = error ->
+        error
+    end
```
Comment (wojtekmach) [inline lib/goth/server.ex:95]:
```suggestion
    with {:ok, token} <- Token.fetch(state) do
      store_and_schedule_refresh(state, token)
      {:ok, token}
    end
```

File: lib/goth/server.ex [hunk 7]
```diff
@@ -21,33 +21,26 @@ defmodule Goth.Server do
...
+      _ -> nil
     end
+  rescue
+    ArgumentError -> nil
   end
 
   @impl true
   def init(opts) when is_list(opts) do
     {backoff_opts, opts} = Keyword.split(opts, [:backoff_type, :backoff_min, :backoff_max])
+    {prefetch, opts} = Keyword.pop(opts, :prefetch)
```
Comment (wojtekmach) [inline lib/goth/server.ex:43]: please document the `:prefetch` option on Goth.start_link/1.

File: lib/goth/server.ex [hunk 8]
```diff
@@ -21,33 +21,26 @@ defmodule Goth.Server do
...
+  def fetch(name) do
+    read_from_ets(name) || GenServer.call(registry_name(name), :fetch)
+  end
 
-      true ->
-        Token.fetch(config)
+  defp read_from_ets(name) do
+    case Registry.lookup(@registry, name) do
+      [{_pid, %Token{} = token}] -> {:ok, token}
+      _ -> nil
```
[the remark this answers is not in the document]
Comment (aleDsz) [reply lib/goth/server.ex:34]: For some reason, with current implementation, is possible to return `[{#PID<0.550.0>, nil}]`

File: lib/goth/server.ex [hunk 9]
```diff
@@ -21,33 +21,26 @@ defmodule Goth.Server do
...
     end
+  rescue
+    ArgumentError -> nil
```
[the remark this answers is not in the document]
Comment (aleDsz) [reply lib/goth/server.ex:37]: I would say we don't need it anymore and we should throw the exception when the user doesn't start the server

**Pull request merged**